src/structures/CommandInteractionOptionResolver.js

```diff
@@ -0,0 +1,159 @@
+'use strict';
+
+/**
+ * A resolver for command interaction options.
+ */
+class CommandInteractionOptionResolver {
+  constructor(options) {
+    /**
+     * The interaction options array.
+     * @type {CommandInteractionOption[]}
+     * @private
+     */
+    this.options = options ?? [];
+  }
+
+  /**
+   * Gets an option by its name.
+   * @param {string} name The name of the option.
+   * @param {boolean} required Whether to throw an error if the option is not found.
+   * @returns {?CommandInteractionOption} The option, if found.
+   */
+  get(name, required = false) {
+    const option = this.options.find(opt => opt.name === name);
+    if (option) {
+      return option;
+    } else if (required) {
+      throw new Error(`Missing required option "${name}"`);
+    } else {
+      return null;
+    }
+  }
+
+  /**
+   * Gets an option by name and property and checks its type.
+   * @param {string} name The name of the option.
+   * @param {ApplicationCommandOptionType[]} types The type of the option.
+   * @param {string[]} properties The property to check for for `required`.
+   * @param {boolean} required Whether to throw an error if the option is not found.
+   * @returns {*|null} The value of `property`.
+   * @private
+   */
+  _getTypedOption(name, types, properties, required) {
+    const option = this.get(name, required);
+    if (option) {
+      if (!types.includes(option.type)) {
+        throw new Error(`Option "${name}" is of type "${option.type}"; expected ${types.join('|')}.`);
+      } else if (required && properties.every(prop => option[prop] === null || typeof option[prop] === 'undefined')) {
+        throw new Error(`Option "${name}" of type "${option.type}" is required, but is empty.`);
+      } else {
+        return option;
+      }
+    } else {
+      return option;
+    }
+  }
+
+  /**
+   * Gets a sub-command or sub-command group.
+   * @param {string} name The name of the sub-command or sub-command group.
+   * @returns {CommandInteractionOptionResolver|null}
+   * A new resolver for the sub-command/group's options, or null if empty
+   */
+  getSubCommand(name) {
+    const option = this._getTypedOption(name, ['SUB_COMMAND', 'SUB_COMMAND_GROUP'], ['options'], false);
```

Comment:
Why is this option type never required?

Reply:
Users will always have at least one subcommand at the same level, but usually more. It doesn't make much sense to me for it to be required.